apps/core/middleware.py

```diff
@@ -1,18 +1,29 @@
-from django.http import Http404
+from django.conf import settings
+from django.urls import LocalePrefixPattern
 from django.utils.translation import activate, get_language
 
 from apps.core.models import HomePage
 
+pattern = LocalePrefixPattern()
+
 
 class ValidateLocaleMiddleware:
     def __init__(self, get_response):
         self.get_response = get_response
 
     def __call__(self, request):
-        try:
-            HomePage.objects.get(locale__language_code=get_language(), live=True)
-        except HomePage.DoesNotExist:
-            activate("en-latest")
-            raise Http404()
-        response = self.get_response(request)
+        if (
+            request.path == "/"
+            or request.resolver_match
+            and pattern.match(request.resolver_match.route)
+        ):
+            try:
+                HomePage.objects.get(locale__language_code=get_language(), live=True)
+                response = self.get_response(request)
+            except HomePage.DoesNotExist:
+                # The requested language is not available, use the default
+                activate(settings.LANGUAGE_CODE)
+                response = self.get_response(request)
+        else:
+            response = self.get_response(request)
```

Comment:
If you call the resolver before `get_response` you will always have None, so this code defaults to the else statement every single time. I ran the tests with a print statement in the else section and it printed every time.

I also tried to set my language to a language that had a matching homepage and it did not properly set my preferred language until I put the `response = self.get_response(request)` at the top.

I see that you removed the 404 response when a language does not exist, was there a reason for that?

Reply:
I'm not sure why the tests do not fail on the language preference, but at least it was not working in the browser for me

Reply:
I think that the tests that correctly pass when a user asks for "/" and is redirected to the preferred language (browser or cookie) because of LocaleMiddleware

Reply:
So, I guess we need to take a step back and actually think about what the requirements are